### ISSUE TITLE
`Trainer.preprocess_dataset` - `KeyError` with `datasets<2.6.0`

### ISSUE THREAD

**Issue by pdhall99** [tomaarsen/SpanMarkerNER#2]
I ran the README snippet in a Kaggle notebook (Python 3.7) after running `%pip install datasets span_marker transformers`. I get a `KeyError` from `trainer.train()` / `Trainer.preprocess_dataset` (see below).

I was able to fix this by upgrading to `datasets>=2.6.0`. It seems that with the default Kaggle setup `datasets==2.1.0` was installed when I ran `%pip install datasets span_marker transformers`.

I can't see an obvious reason for the change in `datasets` behaviour in the 2.6.0 release notes so maybe I'm on the wrong track. I thought you might like to know, in case there is a reason and you would like to add a constraint on `datasets` as a dependency.

Thanks for the package!

Traceback:
```
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
/tmp/ipykernel_103/2689077115.py in <module>
     30 )
     31 
---> 32 trainer.train()
     33 trainer.save_model("my_span_marker_model/checkpoint-final")
     34 

/opt/conda/lib/python3.7/site-packages/transformers/trainer.py in train(self, resume_from_checkpoint, trial, ignore_keys_for_eval, **kwargs)
   1635             resume_from_checkpoint=resume_from_checkpoint,
   1636             trial=trial,
-> 1637             ignore_keys_for_eval=ignore_keys_for_eval,
   1638         )
   1639 

/opt/conda/lib/python3.7/site-packages/transformers/trainer.py in _inner_training_loop(self, batch_size, args, resume_from_checkpoint, trial, ignore_keys_for_eval)
   1643         self._train_batch_size = batch_size
   1644         # Data loader and number of training steps
-> 1645         train_dataloader = self.get_train_dataloader()
   1646 
   1647         # Setting up training control variables:

/opt/conda/lib/python3.7/site-packages/span_marker/trainer.py in get_train_dataloader(self)
    176     def get_train_dataloader(self) -> DataLoader:
    177         """Return the preprocessed training DataLoader."""
--> 178         self.train_dataset = self.preprocess_dataset(self.train_dataset, self.label_normalizer, self.tokenizer)
    179         return super().get_train_dataloader()
    180 

/opt/conda/lib/python3.7/site-packages/span_marker/trainer.py in preprocess_dataset(self, dataset, label_normalizer, tokenizer, dataset_name, is_evaluate)
    170             batched=True,
    171             remove_columns=dataset.column_names,
--> 172             desc=f"Tokenizing the {dataset_name} dataset",
    173         )
    174         return dataset

/opt/conda/lib/python3.7/site-packages/datasets/arrow_dataset.py in map(self, function, with_indices, with_rank, input_columns, batched, batch_size, drop_last_batch, remove_columns, keep_in_memory, load_from_cache_file, cache_file_name, writer_batch_size, features, disable_nullable, fn_kwargs, num_proc, suffix_template, new_fingerprint, desc)
   1971                 new_fingerprint=new_fingerprint,
   1972                 disable_tqdm=disable_tqdm,
-> 1973                 desc=desc,
   1974             )
   1975         else:

/opt/conda/lib/python3.7/site-packages/datasets/arrow_dataset.py in wrapper(*args, **kwargs)
    518             self: "Dataset" = kwargs.pop("self")
    519         # apply actual function
--> 520         out: Union["Dataset", "DatasetDict"] = func(self, *args, **kwargs)
    521         datasets: List["Dataset"] = list(out.values()) if isinstance(out, dict) else [out]
    522         for dataset in datasets:

/opt/conda/lib/python3.7/site-packages/datasets/arrow_dataset.py in wrapper(*args, **kwargs)
    485         }
    486         # apply actual function
--> 487         out: Union["Dataset", "DatasetDict"] = func(self, *args, **kwargs)
    488         datasets: List["Dataset"] = list(out.values()) if isinstance(out, dict) else [out]
    489         # re-apply format to the output

/opt/conda/lib/python3.7/site-packages/datasets/fingerprint.py in wrapper(*args, **kwargs)
    456             # Call actual function
    457 
--> 458             out = func(self, *args, **kwargs)
    459 
    460             # Update fingerprint of in-place transforms + update in-place history of transforms

/opt/conda/lib/python3.7/site-packages/datasets/arrow_dataset.py in _map_single(self, function, with_indices, with_rank, input_columns, batched, batch_size, drop_last_batch, remove_columns, keep_in_memory, load_from_cache_file, cache_file_name, writer_batch_size, features, disable_nullable, fn_kwargs, new_fingerprint, rank, offset, disable_tqdm, desc, cache_only)
   2341                                 indices,
   2342                                 check_same_num_examples=len(input_dataset.list_indexes()) > 0,
-> 2343                                 offset=offset,
   2344                             )
   2345                         except NumExamplesMismatchError:

/opt/conda/lib/python3.7/site-packages/datasets/arrow_dataset.py in apply_function_on_filtered_inputs(inputs, indices, check_same_num_examples, offset)
   2218             if with_rank:
   2219                 additional_args += (rank,)
-> 2220             processed_inputs = function(*fn_args, *additional_args, **fn_kwargs)
   2221             if update_data is None:
   2222                 # Check if the function returns updated examples

/opt/conda/lib/python3.7/site-packages/datasets/arrow_dataset.py in decorated(item, *args, **kwargs)
   1913                 )
   1914                 # Use the LazyDict internally, while mapping the function
-> 1915                 result = f(decorated_item, *args, **kwargs)
   1916                 # Return a standard dict
   1917                 return result.data if isinstance(result, LazyDict) else result

/opt/conda/lib/python3.7/site-packages/span_marker/trainer.py in <lambda>(batch)
    167         # Tokenize and add start/end markers
    168         dataset = dataset.map(
--> 169             lambda batch: tokenizer(batch["tokens"], labels=batch["ner_tags"], return_num_words=is_evaluate),
    170             batched=True,
    171             remove_columns=dataset.column_names,

/opt/conda/lib/python3.7/site-packages/datasets/arrow_dataset.py in __getitem__(self, key)
    123 class Batch(LazyDict):
    124     def __getitem__(self, key):
--> 125         values = super().__getitem__(key)
    126         if self.features and key in self.features:
    127             values = [

/opt/conda/lib/python3.7/collections/__init__.py in __getitem__(self, key)
   1025         if hasattr(self.__class__, "__missing__"):
   1026             return self.__class__.__missing__(self, key)
-> 1027         raise KeyError(key)
   1028     def __setitem__(self, key, item): self.data[key] = item
   1029     def __delitem__(self, key): del self.data[key]

KeyError: 'tokens'
```

**Comment by pdhall99** [comment]
It looks like the issue is caused by differences in whether any columns not in the `input_columns` list in `Dataset.map` are removed in the output or not: 
https://github.com/tomaarsen/SpanMarkerNER/blob/fcfa272670bae05b64eb429a746c8e6bc97f1494/span_marker/trainer.py#L161-L165

Relevant `datasets` PRs:
https://github.com/huggingface/datasets/pull/4971
https://github.com/huggingface/datasets/pull/5006
https://github.com/huggingface/datasets/pull/5008

**Comment by tomaarsen** [comment]
@pdhall99 Thank you for this detailed report! Based on your last comment, it seems like I could update:
https://github.com/tomaarsen/SpanMarkerNER/blob/fcfa272670bae05b64eb429a746c8e6bc97f1494/span_marker/trainer.py#L163
to
```python
    input_columns=("tokens", "ner_tags"),
```
with minor modifications to the label normalizers. 

I'll work on that today! I'll try to get better defined minimum versions for core dependencies like `datasets`, much like how SpanMarker needs `transformers>=4.19.0`.

- Tom Aarsen

**Comment by tomaarsen** [comment]
Solved via 98e38802a85b8903a79598f6fb2b21f2df6f5464. 
I'm afraid it'll be included in the upcoming v1.0.0, which has some breaking changes that make <1.0.0 models perform worse or break.
